File: 6.rs
impl Solution {
    pub fn sort_array(mut nums: Vec<i32>) -> Vec<i32> {
        let n = nums.len();
        let mut buf = vec![0; n];
        Self::merge_sort(&mut nums, &mut buf, 0, n);
        nums
    }

    fn merge_sort(arr: &mut [i32], buf: &mut [i32], start: usize, end: usize) {
        if end - start <= 1 {
            return;
        }

        let mid = start + (end - start) / 2;
        Self::merge_sort(arr, buf, start, mid);
        Self::merge_sort(arr, buf, mid, end);

        Self::merge(arr, buf, start, mid, end);
    }

    fn merge(arr: &mut [i32], buf: &mut [i32], start: usize, mid: usize, end: usize) {
        let mut i = start;
        let mut j = mid;
        let mut k = start;

        while i < mid && j < end {
            if arr[i] <= arr[j] {
                buf[k] = arr[i];
                i += 1;
            } else {
                buf[k] = arr[j];
                j += 1;
            }
            k += 1;
        }

        while i < mid {
            buf[k] = arr[i];
            i += 1;
            k += 1;
        }

        while j < end {
            buf[k] = arr[j];
            j += 1;
            k += 1;
        }

        arr[start..end].copy_from_slice(&buf[start..end]);
    }
}

File: 6.cpp
#include <algorithm>
#include <vector>

using namespace std;

using It = vector<int>::iterator;

static void merge(It l_start, It l_end, It r_start, It r_end, It tbuf) {
  while (l_start != l_end && r_start != r_end)
    iter_swap(tbuf++, (*l_start < *r_start) ? l_start++ : r_start++);
  while (l_start != l_end)
    iter_swap(l_start++, tbuf++);
  while (r_start != r_end)
    iter_swap(r_start++, tbuf++);
}

static void merge_sort(It start, It end);

static void merge_sort_internal(It start, It end, It tbuf) {
  if (start == end)
    return;

  if (next(start) == end) {
    iter_swap(tbuf, start);
    return;
  }

  It mid = start + distance(start, end) / 2;

  merge_sort(start, mid);
  merge_sort(mid, end);

  merge(start, mid, mid, end, tbuf);
}

static void merge_sort(It start, It end) {
  if (start == end || next(start) == end)
    return;

  It mid = start + distance(start, end) / 2;
  It tbuf = end - distance(start, mid);

  merge_sort_internal(start, mid, tbuf);

  while (distance(start, tbuf) > 1) {
    mid = tbuf;
    tbuf = start + (distance(start, mid) + 1) / 2;

    merge_sort_internal(tbuf, mid, start);
    merge(start, start + (distance(tbuf, mid)), mid, end, tbuf);
  }

  if (distance(start, tbuf) == 1) {
    for (It it = start; next(it) != end && *it > *next(it); it++)
      iter_swap(it, next(it));
  }
}

class Solution {
public:
  vector<int> sortArray(vector<int> &nums) {
    merge_sort(nums.begin(), nums.end());
    return nums;
  }
};
